docs/Plugins/Plugin Configuration.md
# Plugin Configuration

You can add customizable behavior to your plugins by specifying configuration options.
These can be specified in one of two ways.

## Static Configuration

When Omejdn is started, it can read YAML files specifying plugins
whose locations are given in the environment variable `OMEJDN_PLUGINS`.
These plugins may come with additional configuration data
available via `PluginLoader.configuration(plugin_name)`
Consider this example `plugins.yml` file:

```yaml
plugins:
  my_awesome_plugin:
    foo: 5
    bar: ["hello"]
```

This loads a plugin from `plugins/my_awesome_plugin/my_awesome_plugin.rb`.
If said plugin calls `PluginLoader.configuration('my_awesome_plugin')`,
it will receive the following hash:

```ruby
{
    'foo' => 5,
    'bar' => ["hello"]
}
```

Static configuration files are only read once and never written to.
If you need to change configuration data,
you might want to consider using dynamic configuration instead
(or in addition to static configuration).

## Dynamic Configuration

Omejdn provides `Config.read_config(section, fallback)` and `Config.write_config(section, data)`,
which you can use to read and write arrays and hashes from and into the configuration store.

By convention, your plugin should use its name as section
and initialize the configuration data with default values right after startup.
This will allow other components to find your configuration data
(Think e.g. about Omejdn's Admin UI).

docs/Configuration/Main Configuration.md
# Omejdn Main Configuration File

Omejdn's main configuration file is located at `config/omejdn.yaml`.
It is a YAML-Formatted configuration file consisting of several configurable options.
It is being reloaded whenever Omejdn needs values from it.
Note however that some changes in values require a restart of Omejdn, such as the loading and unloading of Plugins.

*Note: If you omitt any configuration options, Omejdn will fill in reasonable defaults for you. This also means that any comments in this file will get overwritten*

## Main Configuration Options

### Issuer Identifier

```
issuer: http://example.org
```

Omejdn's OAuth 2.0 Issuer identifier.
It is used as the `iss` value in any issued tokens and advertised as `issuer` in the OAuth Server Metadata.

It is important that, per RFC 8414, the server metadata document is made available
at a URL given by the Issuer Identifier's scheme and authority,
followed by `/.well-known/oauth-authorization-server`, followed by the Issuer Identifier's optional path.
For example, if the Issuer Identifier is `https://example.org/my/path`,
then the URL `https://example.org/.well-known/oauth-authorization-server/my/path` must serve Omejdn's `/.well-known/oauth-authorization-server` endpoint.
For backwards compatibility, the same should be true for the `/.well-known/openid-configuration` infix.
See [NginX](../Integration/NginX.md) for the necessary configuration with NginX.

_Defaults to http://localhost:4567 for testing purposes_

### Front URL

```
issuer: http://example.org/auth
```

The prefix to Omejdn's endpoints.
This is used in several places, the most important one being the OAuth Server Metadata Document,
where e.g. the token endpoint is determined by taking the front URL and appending `/token`.

_Defaults to the `issuer` value_

### Port and Address Binding

```
bind_to: 0.0.0.0/4567
```

An IP address and optional port to bind to.

_Default value: 0.0.0.0/4567_

### Application Environment

```
environment: debug
```
Specifies the level of verbosity for Omejdn.
Can take the following values:

- `production` Do not output debug information
- `development` Be more verbose
- `test` For testing purposes

_Default value: debug_

### Enable OpenID Functionality

```
openid: true
```

This option enables the OpenID connect functionality of Omejdn.
Note that it has to be explicitly enabled.

_Default: false_

### Security Token Settings

```
token:
  expiration: 3600
  algorithm: RS256
id_token:
  expiration: 3600
  algorithm: RS256
```

**Required** Parameters affecting the issued access and ID tokens.
Specified are

- `expiration` The token lifetime in seconds
- `algorithm` The signature algorithm to use.

_Defaults to `3600` for the expiration times and `RS256` for the algorithm_

### Default User Backend

```
user_backend_default: yaml
```

Specifies a plugin to use by default for users created via the Admin API.
Must match one of the enabled User Database Plugins.

_If `openid` is enabled, set to the first specified plugin by default. If not, it is not specified_

### Disable Password Login

```
no_password_login: false
```

Hide the password login when authorizing. Users may still use other Providers. 

_Default: false_

### Default Target Audience

```
default_audience: https://example.org
```

Specifies a default value for the `aud` claim in issued access tokens.
Per RFC 8707, clients may request other values using the `resource` request parameter.
Those other values are granted depending on the client's configuration

_Defaults to ""_

### Accepted Audience Values

```
accept_audience: https://example.org
```

A list of identifiers for which Omejdn is responsible.
When using the `private_key_jwt` or `client_secret_jwt` Client Authentication Methods,
one of the values specified in the `aud` claim has to match one of the values specified here.

_Defaults to the `issuer` value_

## Plugins

Plugins have their own custom configuration, which *should* be documented in the corresponding folder.

In general, the configuration in the configuration file looks like this:

```
plugins:
  user_db:
    yaml:
      location: config/users.yml
  api:
    admin_v1:
    user_selfservice_v1:
      allow_deletion: true
      allow_password_change: true
      editable_attributes:
      - name
  claim_mapper:
    attribute:
```

The Ruby files for the plugins are expected to be found in `plugins/<type>/<name>.rb`,
and they are enabled in the configuration file like this:

```
plugins:
  <type>:
    <name>:
      configuration_option_1: value_1
      configuration_option_2: value_2
```

Different types of plugins are used for different purposes within Omejdn.
Please have a look at the corresponding folder in this documentation.

release_notes.md
# Release Notes

Please read these notes carefully before updating.
**Bold** notes require manual intervention.

## V 1.2.0

- Added support for dynamically chosen claim values
- Added several configuration options to `omejdn.yml`, which were previously only configurable via environment variables.
- Bugfixes and tests

## V 1.1.1

- **The default user backend for new users can (and must) now be specified using the new `user_backend_default` key in `omejdn.yml`**
- Mainly Bugfixes

## V 1.1.0

- **Client certificate files are now chosen automatically. You may import your current certs by specifying a file as `import_certfile` for each client. `certfile` is now depricated**
- **The user selfservice API must now be explicitly enabled in `omejdn.yml`**
- RFC 8707 resource indication capabilities

lib/config.rb
# frozen_string_literal: true

require 'yaml'
CONFIG_SECTION_OMEJDN            = 'omejdn'
CONFIG_SECTION_CLIENTS           = 'clients'
CONFIG_SECTION_USERS             = 'users'
CONFIG_SECTION_OAUTH_PROVIDERS   = 'oauth_providers'
CONFIG_SECTION_SCOPE_DESCRIPTION = 'scope_description'
CONFIG_SECTION_SCOPE_MAPPING     = 'scope_mapping'
CONFIG_SECTION_WEBFINGER         = 'webfinger'

# Configuration helpers functions
class Config
  def self.write_config(section, data)
    PluginLoader.fire('CONFIGURATION_STORE', binding)
  end

  def self.read_config(section, fallback)
    PluginLoader.fire('CONFIGURATION_LOAD', binding).first
  end

  def self.client_config
    read_config CONFIG_SECTION_CLIENTS, []
  end

  def self.client_config=(config)
    write_config(CONFIG_SECTION_CLIENTS, config)
  end

  def self.user_config
    read_config CONFIG_SECTION_USERS, []
  end

  def self.user_config=(config)
    write_config(CONFIG_SECTION_USERS, config)
  end

  def self.base_config
    read_config CONFIG_SECTION_OMEJDN, {}
  end

  def self.base_config=(config)
    write_config CONFIG_SECTION_OMEJDN, config
  end

  def self.oauth_provider_config
    read_config CONFIG_SECTION_OAUTH_PROVIDERS, []
  end

  def self.oauth_provider_config=(providers)
    write_config(CONFIG_SECTION_OAUTH_PROVIDERS, providers)
  end

  def self.scope_description_config
    read_config CONFIG_SECTION_SCOPE_DESCRIPTION, {}
  end

  def self.scope_mapping_config
    read_config CONFIG_SECTION_SCOPE_MAPPING, {}
  end

  def self.webfinger_config
    read_config CONFIG_SECTION_WEBFINGER, {}
  end

  def self.webfinger_config=(config)
    write_config(CONFIG_SECTION_WEBFINGER, config)
  end

  # Fill missing values in the main configuration
  # This will create a configuration file if necessary
  def self.setup
    # Load existing configuration
    config = base_config

    # Fill in default values
    apply_env(config, 'issuer',           'http://localhost:4567')
    apply_env(config, 'front_url',        config['issuer'])
    apply_env(config, 'bind_to',          '0.0.0.0:4567')
    apply_env(config, 'environment',      'development')
    apply_env(config, 'openid',           false)
    apply_env(config, 'default_audience', [])
    apply_env(config, 'accept_audience',  [config['issuer'], "#{config['front_url']}/token"])
    %w[access_token id_token].each do |token|
      apply_env(config, "#{token}.expiration", 3600)
      apply_env(config, "#{token}.algorithm",  'RS256')
    end

    # Save configuration file
    Config.base_config = config
  end

  def self.apply_env(config, conf_key, fallback)
    conf_parts = conf_key.split('.')
    env_value = ENV.fetch("OMEJDN_#{conf_parts.join('__').upcase}", nil)
    conf_key = conf_parts.pop
    conf_parts.each do |part|
      config[part] ||= {}
      config = config[part]
    end
    env_value = env_value.to_i if begin
      Integer(env_value)
    rescue StandardError
      false
    end
    env_value = false if env_value == 'false'
    env_value = true if env_value == 'true'
    config[conf_key] = env_value || config[conf_key] || fallback
  end

  # Initialize admin user if given in ENV
  def self.create_admin
    return unless ENV['OMEJDN_ADMIN']

    admin_name, admin_pw = ENV['OMEJDN_ADMIN'].split(':')
    admin = User.find_by_id(admin_name)
    unless admin
      admin = User.from_h({
                            'username' => admin_name,
                            'attributes' => [{ 'key' => 'omejdn', 'value' => 'admin' }]
                          })
      User.add_user(admin, Config.base_config['user_backend_default'])
    end
    admin.update_password(admin_pw)
  end
end

# DefaultConfigDB saves Configuration Data as YAML files on disk
class DefaultConfigDB
  CONFIG_DIR = 'config'

  def self.write_config(bind)
    section = bind.local_variable_get :section
    data    = bind.local_variable_get :data
    file = File.new "#{CONFIG_DIR}/#{section}.yml", File::CREAT | File::TRUNC | File::RDWR
    file.write data.to_yaml
    file.close
  end

  def self.read_config(bind)
    section  = bind.local_variable_get :section
    fallback = bind.local_variable_get :fallback
    (YAML.safe_load (File.read "#{CONFIG_DIR}/#{section}.yml"), fallback: fallback,
                                                                filename: "#{CONFIG_DIR}/#{section}.yml") || fallback
  end

  # register functions
  def self.register
    PluginLoader.register 'CONFIGURATION_STORE', method(:write_config)
    PluginLoader.register 'CONFIGURATION_LOAD',  method(:read_config)
  end
end
